library/std/src/io/buffered/mod.rs

```diff
@@ -126,6 +126,28 @@ impl<W> IntoInnerError<W> {
     pub fn into_inner(self) -> W {
         self.0
     }
+
+    /// Consumes the [`IntoInnerError`] and returns the error which caused the call to
+    /// [`BufWriter::into_inner()`] to fail, and the underlying writer.
+    ///
+    /// This can be used to simply obtain ownership of the underlying error; it can also be used for
+    /// advanced error recovery.
+    ///
+    /// # Example:
+    /// ```
+    /// use std::io::{BufWriter, ErrorKind, Write};
+    ///
+    /// let mut not_enough_space = [0u8; 10];
+    /// let mut stream = BufWriter::new(not_enough_space.as_mut());
+    /// write!(stream, "this cannot be actually written").unwrap();
+    /// let into_inner_err = stream.into_inner().expect_err("now we discover it's too small");
+    /// let (err, _writer) = into_inner_err.into_parts();
+    /// assert_eq!(err.kind(), ErrorKind::WriteZero);
```

Comment:
Can you think of an example where both parts would be used afterwards?

(This example shows there might be a good reason to (also) add a `.into_error()`.)

Reply:
I would be very happy to add `.into_error()` too.

For a hypothetical situation where one might want the underlying writer too: perhaps the underlying writer is some kind of network connection which can be reset to a known state and then reused.  Eg, maybe the send operation on a TCP socket has timed out and you can recover by advancing the TCP Urgent Data Pointer in the expectation that this will both get the receiver to start discarding data, and then to resynchronise.

Or maybe the underlying writer has interesting state which you want to use for error reporting.  Or resources you wish to reuse.  Consider a writer which encapsulates an ssh Channel within an owned ssh connection - maybe the write failed because the peer closed the Channel, but you can do something sensible if you can get the ssh connection out of the Channel which involves peeling off the layers.

Reply:
Oh I'm sure it'll be useful to have both. I was just wondering if we can come up with a minimal example for in te documentation that shows usage of both the writer and the error.